### PR TITLE
#4078: Instant sidebar loading via browserAction

### PR DESCRIPTION
**Pull request by fregante**
## What does this PR do?

- Part of #4078

## Reviewer Tips

- Before trying this PR, open your PB sidebar to see a slow-responding browser action for the last time [^1]

    [^1]: For best effects, use on hosts without permissions. It does not affect the actual sidebar.html loading time. Contact your doctor if page loads take longer than 4 hours.

## Discussion

- Further readability improvements were possible but I tried to keep this PR slim

### Before

https://user-images.githubusercontent.com/1402241/185910211-271490e8-8b59-473b-8302-aa30eaa12d34.mov

### After

https://user-images.githubusercontent.com/1402241/185910234-83ac4931-8cb8-4380-808b-d2531c95f793.mov

## Future Work

- [ ] Merge `ensureSidebar` into `showSidebar`
- [ ] Extract `showSidebar` side effects out of it, its logic is hard to follow and it's doing two things at once.
- [ ] Rename `hideSidebar`/`showSidebar` with `unloadSidebar`/`loadSidebar` for clarity
- [ ] Drop jQuery from `sidebarDomControllerLite` to make it smaller and avoid issues with it

## Checklist

- [x] Designate a primary reviewer: @twschiller

### CODE REVIEW

File: src/background/browserAction.ts [hunk 1]
```diff
@@ -41,8 +41,17 @@ async function handleBrowserAction(tab: Tabs.Tab): Promise<void> {
...
+      // Toggle the sidebar every time it's run
+      executeScript({
+        tabId: tab.id,
+        files: ["browserActionInstantHandler.js"],
+      }),
+
+      // Run the usual content script unless it's already loaded
+      ensureContentScript({ tabId: tab.id, frameId: TOP_LEVEL_FRAME_ID }),
+    ]);
+    await rehydrateSidebar({
```
Comment (fregante) [inline src/background/browserAction.ts:54]: Since the DOM changes are made by the external `browserActionInstantHandler.js` script, this new method is only used to "confirm" the current state of the UI.

A future logic cleaning around `showSidebar` could make this more clear or merge it into it.

File: src/contentScript/messenger/api.ts [hunk 2]
```diff
@@ -28,12 +28,11 @@ export const resetTab = getNotifier("RESET_TAB");
 
 export const toggleQuickBar = getMethod("TOGGLE_QUICK_BAR");
 export const handleMenuAction = getMethod("HANDLE_MENU_ACTION");
-export const toggleSidebar = getMethod("TOGGLE_SIDEBAR");
+export const rehydrateSidebar = getMethod("REHYDRATE_SIDEBAR");
 export const showSidebar = getMethod("SHOW_SIDEBAR");
 export const hideSidebar = getMethod("HIDE_SIDEBAR");
 export const reloadSidebar = getMethod("RELOAD_SIDEBAR");
 export const removeSidebar = getMethod("REMOVE_SIDEBAR");
-export const getSidebarEntries = getMethod("GET_SIDEBAR_ENTRIES");
```
Comment (fregante) [inline src/contentScript/messenger/api.ts:36]: Not used anywhere

File: src/contentScript/sidebarController.tsx [hunk 3]
```diff
@@ -108,13 +62,16 @@ export async function showSidebar(
...
+  const isShowing = isSidebarFrameVisible();
 
   if (!isShowing) {
-    console.debug("SidePanel is not on the page, attaching side panel");
-    insertSidebar();
+    insertSidebarFrame();
+  }
+
+  try {
     await pingSidebar({ tabId: "this", page: "/sidebar.html" });
```
Comment (fregante) [inline src/contentScript/sidebarController.tsx:72]: New fix: the sidebar is awaited regardless of `isShowing`, or else only the successive `showSidebar` calls would resolve immediately even if the sidebar hasn't loaded.

File: src/contentScript/sidebarController.tsx [hunk 4]
```diff
@@ -108,13 +62,16 @@ export async function showSidebar(
   activateOptions: ActivatePanelOptions = {}
 ): Promise<void> {
   reportEvent("SidePanelShow");
-
-  const isShowing = isSidebarVisible();
+  const isShowing = isSidebarFrameVisible();
```
Comment (fregante) [inline src/contentScript/sidebarController.tsx:65]: I renamed these functions to clarify that they act on the DOM element and do not include any business logic. In the future I'll also adjust showSidebar/hideSidebar to better describe their actions.

File: src/contentScript/sidebarController.tsx [hunk 5]
```diff
@@ -391,7 +336,3 @@ export function upsertPanel(
 
   renderPanelsIfVisible();
 }
-
-if (IS_BROWSER) {
-  storeOriginalCSS();
```
Comment (fregante) [inline src/contentScript/sidebarController.tsx:396]: This is now called only when needed

File: src/contentScript/sidebarDomControllerLite.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+export function isSidebarFrameVisible(): boolean {
+  return Boolean(document.querySelector(PANEL_CONTAINER_SELECTOR));
+}
+
+/** Removes the element; Returns false if no element was found */
+export function removeSidebarFrame(): boolean {
+  if (!isSidebarFrameVisible()) {
+    return false;
```
Comment (fregante) [inline src/contentScript/sidebarDomControllerLite.ts:44]: With the exception of a boolean return and `storeOriginalCSSOnce`, these function have been copied unchanged.

File: src/contentScript/sidebarController.tsx [hunk 7]
```diff
@@ -108,13 +62,16 @@ export async function showSidebar(
...
   if (!isShowing) {
-    console.debug("SidePanel is not on the page, attaching side panel");
-    insertSidebar();
+    insertSidebarFrame();
+  }
+
+  try {
     await pingSidebar({ tabId: "this", page: "/sidebar.html" });
+  } catch (error) {
+    throw new Error("The sidebar did not respond in time", { cause: error });
```
Comment (fregante) [inline src/contentScript/sidebarController.tsx:74]: This wraps the confusing "No handler for PING_SIDEBAR message" like:

- #4058

**Comment by twschiller** [comment]
image
